### PR TITLE
[DirectX][DXIL] Align type spec of TableGen DXIL Op and LLVM Intrinsic

### LINKED ISSUE #86229: [DirectX][DXIL] Align TableGen DXIL Op type spec with that of LLVM Intrinsic
**Problem Description**

Consider an HLSL intrinsic with `void` return type, such as `void barrier(mask)`. Its TableGen specification would be
```
  def int_dx_barrier  : Intrinsic<[], [llvm_i32_ty], [IntrNoDuplicate, IntrWillReturn]>;
```
The specification of DXIL Op that maps to it would be
```
  def Barrier : DXILOpMapping<80, barrier, int_dx_barrier,
                          "Inserts a memory barrier in the shader",
                          [llvm_void_ty, llvm_i32_ty]>;
```
There should be no need to specify overriding types for the DXIL Op `Barrier`, as its types are the same as those of the LLVM Intrinsic it maps to. Such a need arises because the current implementation in `DXIL.td` represents return types and parameter types as a single combined list with an assumption that return type at index `0` of the list. 

**Solution**

The specification of return and parameter types needs to be separated as is done in LLVM Intrinsic specification. As a result, the type specification will be uniform across LLVM Intrinsic and DXIL Op. Consequently, DXIL Op type specification can inherit types from LLVM Intrinsic and can be specified more succinctly.

Changes to `DXILEmitter` to handle the separate specification of return and parameter types also need to be made.

### PR DESCRIPTION
**Pull request by bharadwajy**
Align specification of return type and parameter type fields of DXIL Op mapping with those of TableGan class Intrinsic.

A `void` return type of LLVM Intrinsic is represented as `[]` in its TableGen description record. Currently, a `void` return type of DXIL Operation is represented as `[llvm_void_ty]`. In addition, return and parameter types are recorded as a single list with an understanding that element at index `0` is the return type.

These changes leverage and align DXIL Op type specification with the type specification of the LLVM Intrinsic. As a result, return and parameter types are now specified as two separate lists no longer requiring a different representation for `void` return type. Additionally, type specification would be more succinct yet equally informative for DXIL Op records for which the same LLVM Intrinsics types are also valid.

Added a test to verify lowering of LLVM intrinsic with `void` return.

Barrier intrinsic has a `void` return type. Specification of its DXIL Op can inherit the types of this intrinsic. The test verifies the changes.

Move OverloadKind to DXILABI.h.

Fixes #86229

### CODE REVIEW

**Review by coopp — APPROVED** [review]
Yeah.. This looks like it is now matching Justin's PR.

File: llvm/include/llvm/Support/DXILABI.h
Comment (bharadwajy) [reply llvm/include/llvm/Support/DXILABI.h]: > This change is dup with [a83ed04](https://github.com/llvm/llvm-project/commit/a83ed0423dcafd4fecb15994a5c5c8b85a56f59b) , may need to sync and rebase.

Done.

**Review by bogner — APPROVED** [review]

**Review by llvm-beanz — CHANGES_REQUESTED** [review]
I want to put the breaks on this PR.

We have a document that is documenting this design which is inconsistent with this PR. Before we merge this we should make the documentation consistent. The doc PR is here: https://github.com/llvm/llvm-project/pull/85170.

**Comment by bharadwajy** [comment]
> I want to put the breaks on this PR.
> 
> We have a document that is documenting this design which is inconsistent with this PR. Before we merge this we should make the documentation consistent. The doc PR is here: #85170.

Updated the document 
1. to reflect the changes proposed in this PR. 
2. with details of another option - along with a link to corresponding changes.

Thanks!

**Review by llvm-beanz — COMMENTED** [review]
I'm not going to block this, but I'm still unsure that the documentation is clear and covers all the cases we need. I'll defer to @bogner.

**Comment by bharadwajy** [comment]
> I'm not going to block this, but I'm still unsure that the documentation is clear and covers all the cases we need. I'll defer to @bogner.

I think I prefer the Option 3 (Leverage the existing classification of DXIL Operations) in the updated design document (#85170) compared to the one proposed in this PR for the reasons specified there. I'll explore its current implementation (changeset link in the design doc) a bit more and submit a PR that potentially overrides this one. 

Thanks!

**Comment by bharadwajy** [comment]
Taking a different design direction - see PR #87803. Closing.

**Pull request closed without merging**
